File: src/main/java/fi/quanfoxes/phases/ParserPhase.java
package fi.quanfoxes.phases;

import java.util.ArrayList;
import java.util.List;

import fi.quanfoxes.Bundle;
import fi.quanfoxes.Phase;
import fi.quanfoxes.Status;
import fi.quanfoxes.lexer.Token;
import fi.quanfoxes.parser.Context;
import fi.quanfoxes.parser.Node;
import fi.quanfoxes.parser.Parser;
import fi.quanfoxes.parser.nodes.FunctionNode;
import fi.quanfoxes.parser.nodes.NodeType;
import fi.quanfoxes.parser.nodes.TypeNode;

public class ParserPhase extends Phase {
    public static class Parse {
        private Context context;
        private Node node;

        public Parse(Context context, Node node) {
            this.context = context;
            this.node = node;
        }

        public Context getContext() {
            return context;
        }

        public Node getNode() {
            return node;
        }
    }

    private List<Exception> errors = new ArrayList<>();

    private void members(Node root) {
        Node node = root.first();

        while (node != null) {
            if (node.getNodeType() == NodeType.TYPE_NODE) {
                final TypeNode type = (TypeNode) node;

                async(() -> {
                    try {
                        type.parse();
                    }
                    catch (Exception e) {
                        errors.add(e);
                    }

                    return Status.OK;
                });    

                members(type);
            }

            node = node.next();
        }
    }

    public void functions(Node parent) {
        Node node = parent.first();

        while (node != null) {
            if (node.getNodeType() == NodeType.TYPE_NODE) {
                TypeNode type = (TypeNode)node;
                functions(type);

            } else if (node.getNodeType() == NodeType.FUNCTION_NODE) {
                final FunctionNode function = (FunctionNode)node;

                async(() -> {
                    try {
                        function.parse();
                    } catch (Exception e) {
                        errors.add(e);
                    }

                    return Status.OK;
                });
            }

            node = node.next();
        }
    }

    @Override
    public Status execute(Bundle bundle) {
        final List<Token>[] files = bundle.get("input_file_tokens", null);

        if (files == null) {
            return Status.error("Nothing to parse");
        }

        final Parse[] parses = new Parse[files.length];

        // Form the 'hull' of the code
        for (int i = 0; i < files.length; i++) {
            final int index = i;

            async(() -> {
                List<Token> tokens = files[index];

                Node node = new Node();
                Context context = Parser.initialize();

                try {
                    Parser.parse(node, context, tokens);
                }
                catch (Exception e) {
                    return Status.error(e.getMessage());
                }

                parses[index] = new Parse(context, node);
                
                return Status.OK;
            });
        }

        sync();

        // Parse types, subtypes and their members
        for (int i = 0; i < files.length; i++) {
            final int index = i;

            async(() -> {
                members(parses[index].getNode());
                return Status.OK;
            });
        }

        sync();

        // Parse types, subtypes and their members
        for (int i = 0; i < files.length; i++) {
            final int index = i;

            async(() -> {
                functions(parses[index].getNode());
                return Status.OK;
            });
        }
        
        sync();

        // Merge all parsed files
        Context context = new Context();
        Node root = new Node();

        for (Parse parse : parses) {
            context.merge(parse.getContext());
            root.merge(parse.getNode());
        }

        bundle.put("parse", new Parse(context, root));

        return Status.OK;
    }
}

File: Zigzag/Phases/ParserPhase.cs
using System;
using System.Collections.Generic;

public class Parse
{
	public Context Context { get; private set; }
	public Node Node { get; private set; }

	public Parse(Context context, Node node)
	{
		Context = context;
		Node = node;
	}
}

public class ParserPhase : Phase
{
<<<<<<< HEAD
	private List<Exception> Errors { get; } = new List<Exception>();
=======
	private readonly List<Exception> Errors = new List<Exception>();
>>>>>>> ec8e325... Improved code quality and implemented basic support for operator overloading

	private void ParseMembers(Node root)
	{
		var node = root.First;

		while (node != null)
		{
			if (node.GetNodeType() == NodeType.TYPE_NODE)
			{
				var type = (TypeNode)node;

				Run(() =>
				{
					try
					{
						type.Parse();
					}
					catch (Exception e)
					{
						Errors.Add(e);
					}

					return Status.OK;
				});

				ParseMembers(type);
			}

			node = node.Next;
		}
	}
	public override Status Execute(Bundle bundle)
	{
		if (!bundle.Contains("input_file_tokens"))
		{
			return Status.Error("Nothing to parse");
		}

		var files = bundle.Get<List<Token>[]>("input_file_tokens");
		var parses = new Parse[files.Length];

		// Form the 'hull' of the code
		for (var i = 0; i < files.Length; i++)
		{
			var index = i;

			Run(() =>
			{
				var tokens = files[index];

				var node = new Node();
				var context = Parser.Initialize();

				try
				{
					Parser.Parse(node, context, tokens);
				}
				catch (Exception e)
				{
					return Status.Error(e.Message);
				}

				parses[index] = new Parse(context, node);

				return Status.OK;
			});
		}

		Sync();

		if (Failed)
		{
			return Status.Error(GetTaskErrors());
		}

		// Parse types, subtypes and their members
		for (int i = 0; i < files.Length; i++)
		{
			int index = i;

			Run(() =>
			{
				ParseMembers(parses[index].Node);
				return Status.OK;
			});
		}

		Sync();

		if (Failed)
		{
			return Status.Error(GetTaskErrors());
		}

		// Merge all parsed files
		var context = new Context();
		var root = new Node();

		foreach (var parse in parses)
		{
			context.Merge(parse.Context);
			root.Merge(parse.Node);
		}

		var function = context.GetFunction("run");

		if (function == null)
		{
			return Status.Error("Couldn't find function 'run'");
		}

		function.Overloads[0].Implement(new List<Type>());

		bundle.Put("parse", new Parse(context, root));

		return Status.OK;
	}
}